.github/actions/base_images/action.yml

```
@@ -1,40 +1,91 @@
+        echo "base image hash is ${hash}, python version is ${{ inputs.python_version }}"
+        echo "base_cache_key=${hash}-${{ inputs.python_version }}" >> $GITHUB_OUTPUT
+      shell: bash
+
+    - name: Restore previously cached images
+      id: cache
+      uses: actions/cache/restore@v3
+      with:
+        key: base-images=${{ steps.hash_key.outputs.base_cache_key }}
+        path: base-images.tar.gz
+
+    - name: Extract images if cached
+      if: steps.cache.outputs.cache-hit == 'true'
+      run: |
+        echo "Base Images were in cache"
+        podman load -i base-images.tar.gz
+      shell: bash
+
+    - name: Check for updates on cached images
+      if: steps.cache.outputs.cache-hit == 'true'
+      run: |
+        sudo podman run --rm --privileged multiarch/qemu-user-static --reset -p yes
+        images=()
+        for IMAGE in base pulp-ci-centos9; do
+          for ARCH in arm64 amd64; do
+            echo "Checking if rebuild needed for ${IMAGE}-${ARCH}"
+            podman run --pull=never pulp/${IMAGE}:ci-${ARCH} bash -c "dnf check-upgrade"
+            if [ $? -gt 0 ]; then
+              echo "Rebuild needed"
+              images+=('${IMAGE}-${ARCH}')
+            fi
+          done
+        done
+        if [ ${#images[@]} -eq 0 ]; then
+          echo "No rebuilds needed :)"
+        else
+          echo "BUILD_IMAGES=[$(echo ${images[@]} | sed 's/ /, /g')]" >> $GITHUB_ENV
+        fi
+      shell: bash
+
+    - name: Set images to build on cache miss
+      if: steps.cache.outputs.cache-hit != 'true'
+      run: |
+        echo "BUILD_IMAGES=['base-arm64', 'base-amd64', 'pulp-ci-centos9-arm64', 'pulp-ci-centos9-amd64']" >> $GITHUB_ENV
+        sudo podman run --rm --privileged multiarch/qemu-user-static --reset -p yes
       shell: bash
 
     - name: Build images
+      if: env.BUILD_IMAGES
       run: |
+        images=(${{ join(fromJSON(env.BUILD_IMAGES), ' ') }})
+        echo "Going to build images: ${images[@]}"
         podman version
         buildah version
-        sudo podman run --rm --privileged multiarch/qemu-user-static --reset -p yes
-        for ARCH in arm64 amd64
-        do
-          podman build --platform linux/$ARCH --format docker --file images/Containerfile.core.base --tag pulp/base:${TEMP_BASE_TAG}-${ARCH} --build-arg PYTHON_VERSION=${PYTHON_VERSION} .
-          podman build --platform linux/$ARCH --format docker --file images/pulp_ci_centos/Containerfile --tag pulp/pulp-ci-centos9:${TEMP_BASE_TAG}-${ARCH} --build-arg FROM_TAG=${TEMP_BASE_TAG}-${ARCH} .
+        for image in "${images[@]}"; do
+          echo "Building image ${image}"
+          arch=${image:(-5):5}
+          case $image in
+            base-arm64 | base-amd64)
+              podman build --platform linux/$arch --format docker --file images/Containerfile.core.base --tag pulp/base:ci-$arch --build-arg PYTHON_VERSION=${{ inputs.python_version }} .
```

I know this may seem overly cautious, but shellcheck would suggest it too, for good reasons.
```suggestion
              podman build --platform "linux/${arch}" --format docker --file images/Containerfile.core.base --tag "pulp/base:ci-${arch}" --build-arg PYTHON_VERSION=${{ inputs.python_version }} .
```